Add ApiCompat run to repo

wtgodbe opened the pull request:
Adds an APICompat run between our ref assemblies (Contract) and the corresponding implementation assemblies. CC @dougbu @JunTaoLuo @Pilchie 

Part of https://github.com/aspnet/AspNetCore/pull/17957. We could also add a run in servicing branches that compares against the previous patch build, but that will be a little more custom (and would require us to restore all the packages we shipped in the last patch)

Currently I only have this running on windows

wtgodbe committed:
Add ApiCompat run

wtgodbe committed:
Add baseline files

wtgodbe commented:
https://github.com/aspnet/AspNetCore/pull/17959/commits/759d9ad803e78be5e496a49a862dc43f1e19ddc7 adds the functionality, https://github.com/aspnet/AspNetCore/pull/17959/commits/de0523717c3ea03210fb4ad43cd6d4ac1ba44daf represents the set of errors logged. We can have the run either write to these baseline files, or fail when a new error appears that wasn't previously in the baseline files. There are a few real errors there we should look at and fix first (we probably don't need to care about `CannotChangeAttribute`)

CC @ericstj

wtgodbe commented on src/Components/Components/src/ApiCompatBaseline.txt:
```diff
@@ -0,0 +1,10 @@
+CannotChangeAttribute : Attribute 'System.AttributeUsageAttribute' on 'Microsoft.AspNetCore.Components.LayoutAttribute' changed from '[AttributeUsageAttribute(4, AllowMultiple=false, Inherited=true)]' in the contract to '[AttributeUsageAttribute(AttributeTargets.Class, AllowMultiple=false, Inherited=true)]' in the implementation.
+CannotChangeAttribute : Attribute 'System.AttributeUsageAttribute' on 'Microsoft.AspNetCore.Components.ParameterAttribute' changed from '[AttributeUsageAttribute(128, AllowMultiple=false, Inherited=true)]' in the contract to '[AttributeUsageAttribute(AttributeTargets.Property, AllowMultiple=false, Inherited=true)]' in the implementation.
+CannotChangeAttribute : Attribute 'System.AttributeUsageAttribute' on 'Microsoft.AspNetCore.Components.RouteAttribute' changed from '[AttributeUsageAttribute(4, AllowMultiple=true, Inherited=false)]' in the contract to '[AttributeUsageAttribute(AttributeTargets.Class, AllowMultiple=true, Inherited=false)]' in the implementation.
+ParameterModifiersCannotChange : Custom modifiers on parameter 'renderBatch' on method 'Microsoft.AspNetCore.Components.RenderTree.Renderer.UpdateDisplayAsync(Microsoft.AspNetCore.Components.RenderTree.RenderBatch)' are 'System.Runtime.InteropServices.InAttribute' in the implementation but 'System.Runtime.InteropServices.InAttribute' in the contract.
```
This looks like a bug in the tool

dougbu replied:
`[In]` is defined in a number of different assemblies. Perhaps something in how we build ref/ projects is very slightly different?

In any case, suggest a centralized exclusion for the `[In] != [In]` case.

wtgodbe commented on src/Http/Routing/src/ApiCompatBaseline.txt:
```diff
@@ -0,0 +1,5 @@
+Compat issues with assembly Microsoft.AspNetCore.Routing:
+CannotChangeAttribute : Attribute 'System.AttributeUsageAttribute' on 'Microsoft.AspNetCore.Routing.HostAttribute' changed from '[AttributeUsageAttribute(68, AllowMultiple=false, Inherited=false)]' in the contract to '[AttributeUsageAttribute(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple=false, Inherited=false)]' in the implementation.
+MembersMustExist : Member 'Microsoft.AspNetCore.Routing.Matching.CandidateState.Values.set(Microsoft.AspNetCore.Routing.RouteValueDictionary)' does not exist in the implementation but it does exist in the contract.
```
We should look into this one

Tratcher replied:
Values.Set is internal in the implementation:
https://github.com/aspnet/AspNetCore/blob/138e801e342af561c8f12c8bfd2591d47129db82/src/Http/Routing/src/Matching/CandidateState.cs#L53
But public in the ref assembly:
https://github.com/aspnet/AspNetCore/blob/138e801e342af561c8f12c8bfd2591d47129db82/src/Http/Routing/ref/Microsoft.AspNetCore.Routing.Manual.cs#L129
That seems like a bug in the ref assembly. It should be internal there too and use InternalsVisibleTo.

wtgodbe replied:
@Pilchie does this example refer to your question? The tool picked up a diff between the `src` file and the corresponding `manual.cs` file

Pilchie replied:
This is not conclusive, since it was public on at least one side.  I'd want to see a case where it reports a mismatch in something that is internal on both sides.

dougbu replied:
It's entirely possible we messed a few things up when creating and updating *.Manual.cs files e.g. did the all-important `internal` keyword get deleted accidentally? Suggest checking GenAPI output w/ `--all` because that tool uses the same base infrastructure as ApiCompat.

wtgodbe commented on src/Servers/Kestrel/Core/src/ApiCompatBaseline.txt:
```diff
@@ -0,0 +1,4 @@
+Compat issues with assembly Microsoft.AspNetCore.Server.Kestrel.Core:
+ParameterModifiersCannotChange : Custom modifiers on parameter 'buffer' on method 'Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpParser<TRequestHandler>.Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpParser<TRequestHandler>.ParseRequestLine(TRequestHandler, System.Buffers.ReadOnlySequence<System.Byte>, System.SequencePosition, System.SequencePosition)' are 'System.Runtim ...
```
These look like bugs in the tool

wtgodbe committed:
Only run ApiCompat on windows

Pilchie reviewed (approved):
Thanks, it's great to have this (apart from those bugs in the tool of course).  However, I'm not confident that it actually protects us all the problems we might face.  Specifically, if the internal surface area that is duplicated into our `Manual.cs` files changes, will we get an error so that we know to update the `Manual.cs` as well as the original file, or does the tool only look at public surface area?

wtgodbe commented:
The tool just compares the two assemblies - I _believe_ it looks at internal surface area as well as public, and would therefore tell us if we added something to a `src` file without updating the `manual.cs` files, but I'm not 100% sure - @ericstj can you confirm?

wtgodbe commented:
Hmm, this approach won't work. APICompat depends on the ref assembly being built before the implementation assembly, which isn't always the case for us. Will try adding another pass thru all the projects post-build.

dougbu commented on Directory.Build.targets:
```diff
@@ -115,6 +115,19 @@
 
+  <PropertyGroup Label="APICompat settings">
+    <!-- Don't run APICompat on projects without ref assemblies -->
+    <RunApiCompat Condition="'$(HasReferenceAssembly)' == 'false'">false</RunApiCompat>
```
`$(HasReferenceAssembly)` is never empty (see https://github.com/aspnet/AspNetCore/blob/release/3.1/Directory.Build.targets#L99-L101). All uses of it should be simplified e.g.
```suggestion
    <RunApiCompat Condition=" !$(HasReferenceAssembly) ">false</RunApiCompat>
```

dougbu commented on Directory.Build.targets:
```diff
@@ -115,6 +115,19 @@
+    <RunApiCompat Condition="'$(HasReferenceAssembly)' == 'false'">false</RunApiCompat>
+    <!-- Only run ApiCompat on Windows -->
+    <RunApiCompat Condition="'$(TargetOsName)' != 'win'">false</RunApiCompat>
+    <ContractDir Condition="'$(HasReferenceAssembly)' == 'true'">$(OutputPath.Replace('$(OutDirName)','$(RefOutDirName)'))</ContractDir>
```
```suggestion
    <ContractDir Condition=" $(HasReferenceAssembly) ">$(OutputPath.Replace('$(OutDirName)','$(RefOutDirName)'))</ContractDir>
```

dougbu commented on src/Antiforgery/src/ApiCompatBaseline.txt:
```diff
@@ -0,0 +1 @@
+Total Issues: 0
```
Can we avoid checking in all of these tiny files? E.g. include _one_ `Total Issues: 0` default file plus those containing issues we actually want to ignore.

dougbu commented on src/Components/Components/src/ApiCompatBaseline.txt:
```diff
@@ -0,0 +1,10 @@
+Compat issues with assembly Microsoft.AspNetCore.Components:
+CannotChangeAttribute : Attribute 'System.AttributeUsageAttribute' on 'Microsoft.AspNetCore.Components.BindElementAttribute' changed from '[AttributeUsageAttribute(4, AllowMultiple=true, Inherited=true)]' in the contract to '[AttributeUsageAttribute(AttributeTargets.Class, AllowMultiple=true, Inherited=true)]' in the implementation.
```
Is it possible to add a blanket exclusion for decimal -> `enum` name pseudo-changes? Worst case would be a central list of cases to ignore, probably all involving `[AttributeUsage]` and `[EditorBrowsable]`.

dougbu replied:
My suggestion is to create a single global exclusions file containing the unique lines from the many exclusion files in this PR. That should mean the exact `4` -> `AttributeTargets.Class` equivalents are listed. In particular, set `$(ApiCompatBaseline)` globally for all of the ref/ & src/ project pairs.

That said, adding options directly to `$(ApiCompatArgs)` may be more flexible i.e. there may be interesting options that help out.

dougbu replied:
Ah well. The single baseline file won't work properly and would be a mess to maintain. The closest thing ApiCompat offers is `--exclude-attributes` and that would ignore the attributes completely rather than recognize `4 == AttributeTargets.Class` and so on.

wtgodbe commented:
Maybe I can make it work during the normal pass, these are the projects complaining about the `ref` assembly not being available when building the impl:

```
Microsoft.AspNetCore.Cryptography.Internal.csproj
Microsoft.AspNetCore.Metadata.csproj
Microsoft.AspNetCore.Http.Features.csproj
Microsoft.AspNetCore.DataProtection.Abstractions.csproj
Microsoft.AspNetCore.Components.csproj
Microsoft.AspNetCore.Components.csproj
```

wtgodbe commented:
https://github.com/aspnet/AspNetCore/blob/release/3.1/src/Security/Authorization/Core/src/Microsoft.AspNetCore.Authorization.csproj doesn't hit the error, and looks almost exactly the same as https://github.com/aspnet/AspNetCore/blob/release/3.1/src/Http/Http.Features/src/Microsoft.AspNetCore.Http.Features.csproj, which does

wtgodbe commented:
Oh, some of this might be because projects in the `Components` dir output their artifacts directly into the source tree...

```
  Microsoft.AspNetCore.Components -> F:\workspace\_work\1\s\src\Components\Components\ref\bin\Release\netcoreapp3.1\Microsoft.AspNetCore.Components.dll
```

But others like `Http.Features` don't do that, and still get the same issue. Very confusing. Still looking around.

dougbu commented:
@wtgodbe the ref/ projects are currently built before the corresponding src/ projects _only_ when targeting `netcoreapp3.0`. To relax this and ensure consistent ref/ before src/ builds, remove the second part of the condition at https://github.com/aspnet/AspNetCore/blob/release/3.1/eng/targets/ResolveReferences.targets#L258. This will add the `@(ProjectReference)` more often.

Note: Do not change how `$(_CompileTfmUsingReferenceAssemblies)` is set. Other uses of that property avoid a number of build problems -- as discussed offline.

dougbu committed:
Ensure ref/ projects are always built before corresponding src/ projects
- allows ApiCompat to run as expected

dougbu committed:
!fixup! further cleanup of Directory.Build.targets
- !important! only run ApiCompat in inner builds
- get ref/ assembly path from the relevant project
  - not all projects write into artifacts/...
- make incompatibilities errors (unless overridden)

nits:
- use `$(HasReferenceAssembly)` short form

dougbu committed:
!fixup! Remove unused `$(RefOutDirName)` property

dougbu commented:
@wtgodbe the broken *.Manual.cs file is still broken and I haven't addressed all comments above.

wtgodbe commented:
The same set of projects is still broken, but now with:

> F:\workspace\_work\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19607.3\build\Microsoft.DotNet.ApiCompat.targets(49,5): error : ResolvedMatchingContract item must be specified to run API compat. [F:\workspace\_work\1\s\src\Components\Components\src\Microsoft.AspNetCore.Components.csproj]

Will look in a bit

dougbu committed:
!fixup! add two quick fixes
- ensure `ResolveCustomReferences` target runs consistently
  - `ResolveAssemblyReferences` didn't actually drive execution and it runs too late anyhow
- remove redundant `$(_GenAPICmd)` setting

dougbu commented:
:up::date: to include two more fixes, one a nit the other important

anurse commented:
I'm trying to clear up the state of outstanding servicing PRs. Whats the situation with this? Just holding waiting to fix it? Is this infrastructure-only (sounds like it)? If so, we can slap the `tell-mode` label on it.

dougbu commented:
@anurse to do this right requires Arcade changes from dotnet/arcade#4585 and I haven't quite got that working yet (thought it's close). Moved the PR to 3.1.x because it's not going to be ready for 3.1.3.
